sycl/source/detail/kernel_name_based_cache_t.hpp

```diff
@@ -19,11 +19,39 @@ namespace sycl {
 inline namespace _V1 {
 namespace detail {
 using FastKernelCacheKeyT = std::pair<ur_device_handle_t, ur_context_handle_t>;
-using FastKernelCacheValT =
-    std::tuple<ur_kernel_handle_t, std::mutex *, const KernelArgMask *,
-               ur_program_handle_t>;
+
+struct FastKernelCacheVal {
+  ur_kernel_handle_t MKernelHandle;    /* UR kernel handle pointer. */
+  std::mutex *MMutex;                  /* Mutex guarding this kernel. */
+  const KernelArgMask *MKernelArgMask; /* Eliminated kernel argument mask. */
+  ur_program_handle_t MProgramHandle;  /* UR program handle corresponding to
+                                     this kernel. */
+  const Adapter *MAdapterPtr;          /* We can keep raw pointer to the
+                                     adapter because during 2-stage shutdown
+                                     the kernel cache is destroyed
+                                     deliberately before the adapter. */
+
+  FastKernelCacheVal(ur_kernel_handle_t KernelHandle, std::mutex *Mutex,
+                     const KernelArgMask *KernelArgMask,
+                     ur_program_handle_t ProgramHandle,
+                     const Adapter *AdapterPtr)
+      : MKernelHandle(KernelHandle), MMutex(Mutex),
+        MKernelArgMask(KernelArgMask), MProgramHandle(ProgramHandle),
+        MAdapterPtr(AdapterPtr) {}
+
+  ~FastKernelCacheVal() {
+    if (MKernelHandle)
+      MAdapterPtr->call<sycl::detail::UrApiKind::urKernelRelease>(
```

Comment:
I think we need to delete the copy constructor and copy-assignment operator.

Reply:
Thanks, done.